hoot-core/src/main/cpp/hoot/core/io/OsmGbdxJsonWriter.cpp

```diff
@@ -0,0 +1,391 @@
+ * copyrights will be updated automatically.
+ *
+ * @copyright Copyright (C) 2017, 2018 DigitalGlobe (http://www.digitalglobe.com/)
+ */
+#include "OsmGbdxJsonWriter.h"
+
+// Boost
+using namespace boost;
+
+// Hoot
+#include <hoot/core/OsmMap.h>
+#include <hoot/core/elements/ElementData.h>
+#include <hoot/core/elements/ElementType.h>
+#include <hoot/core/elements/Node.h>
+#include <hoot/core/elements/Relation.h>
+#include <hoot/core/elements/Tags.h>
+#include <hoot/core/elements/Way.h>
+#include <hoot/core/filters/NoInformationCriterion.h>
+#include <hoot/core/index/OsmMapIndex.h>
+#include <hoot/core/schema/OsmSchema.h>
+#include <hoot/core/util/ConfigOptions.h>
+#include <hoot/core/util/Exception.h>
+#include <hoot/core/util/Factory.h>
+#include <hoot/core/util/MetadataTags.h>
+#include <hoot/core/util/OsmUtils.h>
+#include <hoot/core/util/UuidHelper.h>
+#include <hoot/core/visitors/CalculateMapBoundsVisitor.h>
+
+// Qt
+#include <QBuffer>
+#include <QDateTime>
+#include <QFile>
+#include <QXmlStreamWriter>
+#include <QtCore/QStringBuilder>
+
+using namespace geos::geom;
+using namespace std;
+
+namespace hoot
+{
+
+HOOT_FACTORY_REGISTER(OsmMapWriter, OsmGbdxJsonWriter)
+
+OsmGbdxJsonWriter::OsmGbdxJsonWriter(int precision)
+  : OsmJsonWriter(precision)
+{
+  _writeHootFormat = false;
+}
+
+void OsmGbdxJsonWriter::open(QString path)
+{
+  if (path.toLower().endsWith(".gbdx"))
+  {
+    path.remove(path.size() - 5, path.size());
+  }
+
+  _outputDir = QDir(path);
+  _outputDir.makeAbsolute();
+
+  if (_outputDir.exists() == false)
+  {
+    if (QDir().mkpath(_outputDir.path()) == false)
+    {
+      throw HootException("Error creating directory for writing.");
+    }
+  }
+}
+
+void OsmGbdxJsonWriter::_openFile()
+{
+  // Close the old file and open a new one
+  if (_fp.isOpen())
+  {
+    _fp.close();
+  }
+
+  QString url = _outputDir.filePath(UuidHelper::createUuid().toString().replace("{", "").replace("}", "") + ".json");
+
+  _fp.setFileName(url);
+
+  if (!_fp.open(QIODevice::WriteOnly | QIODevice::Text))
+  {
+    throw Exception(QObject::tr("Error opening %1 for writing").arg(url));
+  }
+
+  _out = &_fp;
+}
+
+void OsmGbdxJsonWriter::write(ConstOsmMapPtr map)
+{
+  _map = map;
+
+  _writeNodes();
+
+  _writeWays();
+
+  _writeRelations();
+
+  // Close the last file.
+  _fp.close();
+}
+
+void OsmGbdxJsonWriter::_writeGeometry(ConstNodePtr n)
+{
+  vector<long> node;
+  node.push_back(n->getId());
+  _writeGeometry(node, "Point");
+}
+
+void OsmGbdxJsonWriter::_writeGeometry(ConstWayPtr w)
+{
+  const vector<long>& nodes = w->getNodeIds();
+  bool isPolygon = OsmSchema::getInstance().isArea(w) ||
+                   nodes[0] == nodes[nodes.size() - 1];
+  _writeGeometry(nodes, (isPolygon) ? "Polygon" : "LineString");
+}
+
+void OsmGbdxJsonWriter::_writeGeometry(ConstRelationPtr r)
+{
+  vector<RelationData::Entry> members = r->getMembers();
+  _writeKvp("type", "GeometryCollection"); _write(",");
```

Comment:
The c++ curmudgeon says: don't put two statements like this on the same line!

Reply:
Noted and fixed.

See lines 184-193 in OsmJsonWriter.cpp